meta/Decays.m

```diff
@@ -0,0 +1,2382 @@
+
+      concreteFieldOnEdgeBetweenVerticesLocal = concreteFieldOnEdgeBetweenVertices;
+      temp = SortBy[Reverse /@ fieldNumberOnEdgeBetweenVertices, Last];
+
+      For[i = 1, i <= Length[temp], i++,
+
+         If[GetFeynArtsTypeName[concreteFieldOnEdgeBetweenVerticesLocal[[i,2]]] === (temp[[i,1]] /. fieldTypes),
+
+            temp[[i]] = temp[[i]] /. If[i==2 || i ==3, concreteFieldOnEdgeBetweenVerticesLocal[[i,1]] -> AntiField[concreteFieldOnEdgeBetweenVerticesLocal[[i,2]]], concreteFieldOnEdgeBetweenVerticesLocal[[i]]];
+            temp[[i]] = temp[[i]] /. (Reverse@concreteFieldOnEdgeBetweenVerticesLocal[[i,1]] -> AntiField[concreteFieldOnEdgeBetweenVerticesLocal[[i,2]]]),
+
+            concreteFieldOnEdgeBetweenVerticesLocal[[{i, i+1}]] = concreteFieldOnEdgeBetweenVerticesLocal[[{i+1, i}]];
+            i=i-1;
+          ]
+      ];
+
+      temp
+];
+
+(* map topology to FeynArts name *)
+FeynArtsTopologyName[topology_] :=
+    Switch[topology,
+      {{0, 0, 0, 1, 0, 0}, {0, 0, 0, 0, 1, 0}, {0, 0, 0, 0, 0, 1}, {1, 0, 0,
+        0, 1, 1}, {0, 1, 0, 1, 0, 1}, {0, 0, 1, 1, 1, 0}}, "T1",
+      {{0, 0, 0, 1, 0}, {0, 0, 0, 1, 0}, {0, 0, 0, 0, 1}, {1, 1, 0, 0,
+        1}, {0, 0, 1, 1, 1}}, "T2",
+      {{0, 0, 0, 1, 0}, {0, 0, 0, 0, 1}, {0, 0, 0, 1, 0}, {1, 0, 1, 0,
+        1}, {0, 1, 0, 1, 1}}, "T3",
+      {{0, 0, 0, 1, 0}, {0, 0, 0, 0, 1}, {0, 0, 0, 0, 1}, {1, 0, 0, 0,
+        2}, {0, 1, 1, 2, 0}}, "T4",
+      {{0, 0, 0, 1, 0}, {0, 0, 0, 0, 1}, {0, 0, 0, 0, 1}, {1, 0, 0, 1,
+        1}, {0, 1, 1, 1, 0}}, "T5",
+      {{0, 0, 0, 1, 0}, {0, 0, 0, 0, 1}, {0, 0, 0, 1, 0}, {1, 0, 1, 0,
+        2}, {0, 1, 0, 2, 0}}, "T6",
+      {{0, 0, 0, 1, 0}, {0, 0, 0, 1, 0}, {0, 0, 0, 0, 1}, {1, 1, 0, 0,
+        2}, {0, 0, 1, 2, 0}}, "T7",
+      {{0, 0, 0, 1, 0, 0}, {0, 0, 0, 1, 0, 0}, {0, 0, 0, 0, 1, 0}, {1, 1, 0,
+        0, 0, 1}, {0, 0, 1, 0, 0, 2}, {0, 0, 0, 1, 2, 0}}, "T8",
+      {{0, 0, 0, 1, 0, 0}, {0, 0, 0, 0, 1, 0}, {0, 0, 0, 1, 0, 0}, {1, 0, 1,
+        0, 0, 1}, {0, 1, 0, 0, 0, 2}, {0, 0, 0, 1, 2, 0}}, "T9",
+      {{0, 0, 0, 1, 0, 0}, {0, 0, 0, 0, 1, 0}, {0, 0, 0, 0, 1, 0}, {1, 0, 0,
+        0, 0, 2}, {0, 1, 1, 0, 0, 1}, {0, 0, 0, 2, 1, 0}}, "T10",
+      _, Print["Error: Cannot map topology to FeynArts name"]; Quit[1]
+    ];
+
+WrapCodeInLoop[indices_, code_] :=
+   (
+      "\n// loops over vertices' indices\n" <>
+      StringJoin@@(MapIndexed[
+      Nest[
+         TextFormatting`IndentText,
+         "for(const auto& index" <> ToString@#1 <> ": index_range<vertex" <> ToString@#1 <> ">()) {\n" <>
+            If[First@#2===Length[indices], "\n" <> TextFormatting`IndentText[code], ""],
+         First@#2 -1
+      ]&,
+      indices
+   ]) <>
+     StringJoin@@ (Nest[
+         TextFormatting`IndentText,
+         "}\n",
+         #-1
+         ]& /@ Reverse@Range@Length[indices]));
+
+WrapCodeInLoopOverInternalVertices[decay_, topology_, diagram_] :=
```

Comment:
This function generates a very longish C++ function. Since this C++ function is so long and many local variables with unique names are generated, I'd propose to introduce local blocks to limit the scopes of the local variables.

Currently the following structure is generated:
~~~.cpp
   // topology T1

   using vertexId54280 = Vertex<Glu, typename bar<Fd>::type, Sd>;
   using vertexId54281 = Vertex<Fd, Glu, typename conj<Sd>::type>;
   using vertexId54282 = Vertex<Ah, typename conj<Sd>::type, Sd>;
   constexpr double symmetryFac54279 {1.000000000000000};
   constexpr double colorFac54278 {2.000000000000000};

   // loops over vertices' indices
   for(const auto& indexId54280: index_range<vertexId54280>()) {
      ...
   }

   // topology T2
   // same structure again with more local variables
~~~
I'd propose to put those structures into local blocks, like so:
~~~.cpp
   // topology T1
   {
      using vertex_1 = Vertex<Glu, typename bar<Fd>::type, Sd>;
      using vertex_2 = Vertex<Fd, Glu, typename conj<Sd>::type>;
      using vertex_3 = Vertex<Ah, typename conj<Sd>::type, Sd>;
      constexpr double symmetry_fac = 1.000000000000000;
      constexpr double color_fac = 2.000000000000000;

      // loops over vertices' indices
      for (...) {
         ...
      }
   }

   // topology T2
   // same structure again with more local variables
   {
      ...
   }
~~~
This allows to have very readable local variable/type names. (I also think we're helping the compiler a bit by restricting the scope of the variables/types.)

Reply:
Done in e96af7e953d689e63adde8725197fcfaf26087a1